panic-probe/src/lib.rs

```diff
@@ -0,0 +1,151 @@
+
+    #[cfg(not(any(feature = "print-rtt", feature = "print-defmt")))]
+    fn print(_: &core::panic::PanicInfo) {}
+
+    #[panic_handler]
+    fn panic(info: &PanicInfo) -> ! {
+        static PANICKED: AtomicBool = AtomicBool::new(false);
+
+        // Guard against infinite recursion, just in case.
+        if PANICKED.load(Ordering::Relaxed) {
+            loop {
+                asm::bkpt();
+            }
+        }
+
+        PANICKED.store(true, Ordering::Relaxed);
+
+        print(info);
+
+        // Trigger a `HardFault` via `udf` instruction.
+
+        // If `UsageFault` is enabled, we disable that first, since otherwise `udf` will cause that
+        // exception instead of `HardFault`.
+        #[cfg(not(any(armv6m, armv8m_base)))]
+        {
+            const SHCSR: *mut u32 = 0xE000ED24usize as _;
+            const USGFAULTENA: usize = 18;
+
+            unsafe {
+                let mut shcsr = core::ptr::read_volatile(SHCSR);
+                shcsr &= !(1 << USGFAULTENA);
+                core::ptr::write_volatile(SHCSR, shcsr);
+            }
+        }
+
+        asm::udf();
+    }
+
+    #[exception]
+    fn HardFault(_: &ExceptionFrame) -> ! {
+        loop {
+            // Make `probe-run` print the backtrace and exit.
+            asm::bkpt();
+        }
+    }
+}
+
+#[cfg(feature = "print-rtt")]
+mod print_rtt {
+    use core::panic::PanicInfo;
+    use rtt_target::rprintln;
+
+    pub fn print(info: &PanicInfo) {
+        rprintln!("{}", info);
+    }
+}
+
+#[cfg(feature = "print-defmt")]
+mod print_defmt {
+    use core::{
+        cmp,
+        fmt::{self, Write},
+        mem,
+        panic::PanicInfo,
+        str,
+    };
+
+    const DEFMT_BUF_SIZE: usize = 128;
+    const OVERFLOW_MARK: &str = "…";
+
+    struct Sink<'a> {
+        buf: &'a mut [u8],
+        pos: usize,
+        overflowed: bool,
+    }
+
+    impl<'a> fmt::Write for Sink<'a> {
+        fn write_str(&mut self, s: &str) -> fmt::Result {
+            if self.overflowed {
+                return Ok(());
+            }
+
+            let buf = mem::replace(&mut self.buf, &mut []);
+            let buf_unused = &mut buf[self.pos..];
+
+            if buf_unused.len() < s.len() {
+                self.overflowed = true;
+            }
+
+            let lim = cmp::min(buf_unused.len(), s.len());
+            buf_unused[..lim].copy_from_slice(&s.as_bytes()[..lim]);
+            self.buf = buf;
+            self.pos += lim;
+            Ok(())
+        }
+    }
+
+    pub fn print(info: &PanicInfo) {
+        let mut buf = [0; DEFMT_BUF_SIZE];
+        let mut sink = Sink {
+            buf: &mut buf,
+            pos: 0,
+            overflowed: false,
+        };
+        write!(sink, "{}", info).ok();
+
+        let msg = str::from_utf8(&sink.buf[..sink.pos]).unwrap_or("<utf-8 error>");
+        let overflow = if sink.overflowed { OVERFLOW_MARK } else { "" };
+        defmt::error!("panic-probe: {:str}{:str}", msg, overflow);
```

Comment:
in this case it doesn't hurt to include "panic-probe" (it doesn't cost anything) but it will become a bit redundant once we add file locations to log frames

Reply:
(removed the prefix, the text already includes "panicked at", so it's clear that a panic has occurred at that point)